libraries/llvm/buffer.c

```diff
@@ -0,0 +1,141 @@
+#ifndef EFFEKT_BUFFER_C
+#define EFFEKT_BUFFER_C
+
+
+// TODO It may be performance-advantageous to implement this C file's semantics
+// in LLVM, since the linker cannot realistically be asked to satisfactorily
+// inline.
+
+
+// Eight bytes for the reference counter.
+// E.g. capacity could also be stored left of the data.
+#define BUFFER_METADATA_WIDTH (8)
+
+uint64_t c_buffer_length(const struct Pos buffer) {
+    return buffer.tag;
+}
+
+// NOTE: This assumes a homogenously used byte order.
+uint64_t *c_buffer_refcount(const struct Pos buffer) {
+    return (uint64_t *) (buffer.obj - 8);
+}
+
+uint8_t *c_buffer_bytes(const struct Pos buffer) {
+    return buffer.obj;
+}
+
+
+struct Pos c_buffer_construct(const uint64_t n, const uint8_t *data) {
+    uint8_t *buf = malloc(BUFFER_METADATA_WIDTH + n * sizeof *buf);
```

Comment:
Woah, you can use `buf` on the right-hand-side of its definition, didn't know.

Reply:
Technically, you can use a syntactically finished declaration as an argument to the `sizeof` operator which maps from the type to the term level. It so happens that this definition's first part is said declaration.